### DISCUSSION TITLE
Covariant Returns and Implicit Interface Implementations

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#3946] [General]
This is currently not support due to a breaking change.

> This is technically a breaking change, as the program below prints "C1.M" today, but would print "C2.M" under the proposed revision.
> ``` c#
> using System;
> interface I1 { object M(); }
> class C1 : I1 { public object M() { return "C1.M"; } }
> class C2 : C1, I1 { public new string M() { return "C2.M"; } }
> class Program
> {
>     static void Main()
>     {
>         I1 i = new C2();
>         Console.WriteLine(i.M());
>     }
> }
> ```
> Due to this breaking change, we might consider not supporting covariant return types on implicit implementations.

Can we allow it with some modifier on declaration site so that it's only an opt-in feature?

```cs
interface I1 { [PermitCovariantReturn] object M(); }
```

**Comment by YairHalberstadt** [comment]
If you're going to have to explicitly opt in, explicit implementations seem like they solve the problem reasonably well.

**Reply by alrz** [reply]
That is only on the declaration site and is applied only once. Explicit implementations must be done in every implementer and is too much boilerplate to manage for this case.

**Reply by alrz** [reply]
For context, this is what I'm trying to do.
```cs
public interface IChild { /* a subset of properties */ }
class Child : IChild { /* actual properties */ }
public interface IParent { IChild Child { get; } }
class Parent : IParent {
    public Child Child { get; set; }
}
```
Using interfaces to define a limited public contract to internal types.

**Reply by YairHalberstadt** [reply]
Source generators seem like they could do this for you. They could detect `[PermitCovariantReturn]` and generate the explicit implementations for you.

**Reply by alrz** [reply]
That is definitely an option. Though I see this as exposing a runtime feature currently unavailable in language. We don't really need a double invocation because runtime supports this directly.

**Reply by YairHalberstadt** [reply]
I doubt the double invocation will have much impact. Since it's to a non virtual instance member, if it's small it will be inlined, and if long, the cost of the invocation will be negligible.

**Reply by YairHalberstadt** [reply]
> Though I see this as exposing a runtime feature currently unavailable in language.

I think it would be really nice if this just works out the box. I'm just not a fan of an opt in only magic attribute
